pennylane/gradients/param_shift_hessian.py

```diff
@@ -16,73 +16,105 @@
-    r"""Generate a list of tapes where the corresponding trainable parameter
-    indices have been shifted by the values given.
+    return diag_recipes, partial_offdiag_recipes
 
-    Args:
-        tape (.QuantumTape): input quantum tape
-        idx (list[int]): trainable parameter indices to shift the parameters of
-        shifts (list[list[float or int]]): nested list of shift values, each
-            list containing a value for each index
 
-    Returns:
-        list[QuantumTape]: List of quantum tapes. Each tape has multiple parameters
-            (indicated by ``idx``) shifted by the values of ``shifts``. The length
-            of the returned list of tapes will match the length of ``shifts``.
-    """
-    params = list(tape.get_parameters())
-    tapes = []
+def _generate_off_diag_tapes(tape, idx, recipe_i, recipe_j):
+    r"""Combine two univariate first order recipes and create
+    multi-shifted tapes to compute the off-diagonal entry of the Hessian."""
 
-    for shift in shifts:
-        new_params = params.copy()
-        shifted_tape = tape.copy(copy_operations=True)
+    if recipe_j[0] is None:
+        return [], None, None
 
-        for id_, s in zip(idx, shift):
-            dtype = new_params[id_].dtype
-            new_params[id_] = new_params[id_] + qml.math.convert_like(s, new_params[id_])
-            new_params[id_] = qml.math.cast(new_params[id_], dtype)
+    # The rows of combined_rulesT contain the coefficients (1), the multipliers (2) and the
+    # shifts (2) in that order, with the number in brackets indicating the number of columns
+    combined_rulesT = _combine_shift_rules_with_multipliers(
+        [qml.math.stack(recipe_i).T, qml.math.stack(recipe_j).T]
+    )
+    # If there are unmultiplied, unshifted tapes, the coefficient is memorized and the term
+    # removed from the list of tapes to create
+    if np.allclose(combined_rulesT[1:3, 0], 1.0) and np.allclose(combined_rulesT[3:5, 0], 0.0):
+        unshifted_coeff = combined_rulesT[0, 0]
+        combined_rulesT = combined_rulesT[:, 1:]
+    else:
+        unshifted_coeff = None
 
-        shifted_tape.set_parameters(new_params)
-        tapes.append(shifted_tape)
+    h_tapes = generate_multishifted_tapes(tape, idx, combined_rulesT[3:5].T, combined_rulesT[1:3].T)
 
-    return tapes
+    return h_tapes, combined_rulesT[0], unshifted_coeff
 
 
-def compute_hessian_tapes(tape, diff_methods, f0=None):
+def expval_hessian_param_shift(
+    tape, argnum, diff_methods, diagonal_shifts, off_diagonal_shifts, f0
+):
     r"""Generate the Hessian tapes that are used in the computation of the second derivative of a
     quantum tape, using analytical parameter-shift rules to do so exactly. Also define a
     post-processing function to combine the results of evaluating the Hessian tapes.
 
     Args:
-        tape (.QuantumTape): input quantum tape
+        tape (pennylane.QNode or .QuantumTape): quantum tape or QNode to differentiate
```

Comment:
I don't think this is true. At this point, the input is always a tape I believe.
```suggestion
        tape (.QuantumTape): quantum tape to differentiate
```